### PR TITLE
feature: Handle v1.1.0 collection JSON schema

### LINKED ISSUE #220: Display collection names on the UI
## Use Case

Add a display name for datasets or collections that can be shown on the UI.

## Acceptance Criteria

_Please describe how you know this is done_

## Details

- Does a collection's name for a dataset override its manifest name?

### LINKED ISSUE #248: Handle collection JSON objects (as well as arrays)
## Use Case

Add compatibility to new collection JSON object schema, as described in `colorizer-data` https://github.com/allen-cell-animated/colorizer-data/issues/35

## Acceptance Criteria

- [ ] Collection loader should handle arrays and object collections data, converting to the object format as needed.
- [ ] Collection loader should load any metadata from the new collection JSON format.
- [ ] Add unit tests for both the old collection data and the new format.

## Details

_Please provide any helpful specifications_

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
Closes #248 and closes #220! Updates the collection loader to handle the new v1.1.0 collection JSON schema, as described in https://github.com/allen-cell-animated/colorizer-data/pull/37, and shows the collection names (if provided) on the interface.

*Estimated size: small, 20 minutes*

Solution
========
- Adds a `collection_utils` file similar to the one that exists for datasets, which updates older legacy JSON files and updates them to the current format.
- Adds unit tests for the new collection and dataset JSON files.

## Type of change
Please delete options that are not relevant.

* New feature (non-breaking change which adds functionality)

## Testing
1. Open this preview link and open the developer console: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-270/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Ftest_v1.1.0%2Fcollection.json
2. You should see the name of the collection in the top header, and the collection metadata should include additional information about the collection.

Displayed name:
image

Collection metadata:
image

### CODE REVIEW

File: src/colorizer/Collection.ts [hunk 1]
```diff
@@ -1,19 +1,14 @@
...
- * Dataset properties in a collection. Collections are defined as .json files containing an array of objects
- * with the following properties:
- *
- * - `path`: a relative path from the base directory of the collection URL, or a URL to a dataset.
- * - `name`: the display name for the dataset.
- */
-export type CollectionEntry = {
-  path: string;
-  name: string;
-};
```
Comment (ShrimpCryptid) [inline src/colorizer/Collection.ts:16]: Moved to `collection_utils.ts`

**Review by interim17 — APPROVED** [review]
Thanks for adding the testing section for making review easiser!

File: tests/Collection.test.ts [hunk 2]
```diff
@@ -44,33 +45,94 @@ describe("Collection", () => {
...
+    ];
+
+    for (const { version, collectionJson } of collectionFormats) {
+      describe(version, () => {
+        it("collects datasets", async () => {
+          const url = "https://e.com/collection.json";
+          const mockFetch = makeMockFetchMethod(url, collectionJson);
+          const collection = await Collection.loadCollection(url, mockFetch);
+
+          expect(collection.getDatasetKeys().length).to.equal(5);
```
Comment (meganrm) [inline tests/Collection.test.ts:83]: small thing: but couldn't you do:
```suggestion
          expect(collection.getDatasetKeys().length).to.equal(datasets.length);
```
?
just in case you add more data to your mock data

File: tests/Collection.test.ts [hunk 3]
```diff
@@ -44,33 +45,94 @@ describe("Collection", () => {
...
+        });
+
+        it("retrieves metadata", async () => {
+          const url = "https://e.com/collection.json";
+          const mockFetch = makeMockFetchMethod(url, collectionJson);
+          const collection = await Collection.loadCollection(url, mockFetch);
+
+          if (semver.lte(version, "v0.0.0")) {
+            expect(collection.metadata).deep.equals({});
+            return;
```
Comment (meganrm) [inline tests/Collection.test.ts:114]: this is a personal preference, but I would have made these two different tests: one that sends in v0.0.0 data and checks for the empty object. and one that sends in v1.1.0 data and has everything below. and the description would be "it returns an empty object if the version is 0.0.0" since it's not really retrieving any metadata in this case. Also is there a possibility of there not being a version or do you add it somewhere (I'm starting here in the review)

File: src/colorizer/Collection.ts [hunk 4]
```diff
@@ -224,16 +222,22 @@ export default class Collection {
...
+      const json = await response.json();
+      collection = updateCollectionVersion(json);
+    } catch (e) {
+      throw new Error(`Could not parse collections JSON data from url '${absoluteCollectionUrl}': '${e}'`);
     }
 
     // Convert JSON array into map
+    if (collection.datasets.length === 0) {
+      throw new Error(
+        `Could not retrieve collections JSON data from url '${absoluteCollectionUrl}': No datasets found.`
```
Comment (meganrm) [inline src/colorizer/Collection.ts:236]: do your tests cover this path?

**Review by meganrm — APPROVED** [review]
Overall looks good. Since you wrote tests, I have suggestions for some improvements in the current tests. It also seems like you're only testing for cases where everything goes well, so maybe think about some edge cases, or at least catching the errors you know you're going to throw. I don't know the data well enough to suggest more edge cases myself

File: tests/Collection.test.ts (hunk 3 above)
Comment (ShrimpCryptid) [reply tests/Collection.test.ts:114]: The versions are defined in a big array up above, so here it's checking versions of the data that I've explicitly labeled as `v0.0.0`, `v1.0.0`, and `v1.1.0`. While I think it makes sense to split these into different tests as the tests are now, I'm also anticipating that more backwards-compatible versions will get added in the future. Structuring this test as it is now (where it progressively tests more and more fields) makes sense to me for that specific use case.

File: src/colorizer/Collection.ts (hunk 4 above)
Comment (ShrimpCryptid) [reply src/colorizer/Collection.ts:236]: Good catch, they don't! Adding that now.

File: tests/Collection.test.ts (hunk 3 above)
Comment (ShrimpCryptid) [reply tests/Collection.test.ts:114]: Mm, taking another look at it, I think it makes sense to split these up for now. (or at least, to make `v0.0.0` its own case!)
